jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,170 @@
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingNestedCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link Content.Sink} backed by another {@link Content.Sink}.
+ * Any content written to this {@link Content.Sink} is buffered,
+ * then written to the delegate's
+ * {@link Content.Sink#write(boolean, ByteBuffer, Callback)}. </p>
+ */
+public class BufferedContentSink implements Content.Sink
+{
+    private final Content.Sink _delegate;
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxBufferSize;
+    private CountingByteBufferAccumulator _accumulator;
+    private boolean _firstWrite = true;
+    private boolean _lastWritten;
+
+    public BufferedContentSink(Content.Sink delegate, ByteBufferPool bufferPool, boolean direct, int maxBufferSize)
+    {
+        if (maxBufferSize <= 0)
+            throw new IllegalArgumentException("maxBufferSize must be > 0, was: " + maxBufferSize);
+        _delegate = delegate;
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+    {
+        if (_lastWritten)
+        {
+            callback.failed(new IOException("complete"));
+            return;
+        }
+        if (_firstWrite)
+        {
+            _accumulator = new CountingByteBufferAccumulator(_bufferPool, _direct, _maxBufferSize);
+            _firstWrite = false;
+        }
+        _lastWritten |= last;
+
+        ByteBuffer current = byteBuffer != null ? byteBuffer : BufferUtil.EMPTY_BUFFER;
+        IteratingNestedCallback writer = new IteratingNestedCallback(callback)
+        {
+            private boolean complete;
+
+            @Override
+            protected Action process()
+            {
+                if (complete)
+                    return Action.SUCCEEDED;
+                boolean write = _accumulator.copyBuffer(current);
+                complete = last && !current.hasRemaining();
+                if (write || complete)
+                {
+                    RetainableByteBuffer buffer = _accumulator.takeRetainableByteBuffer();
+                    _delegate.write(complete, buffer.getByteBuffer(), Callback.from(this, buffer::release));
+                    return Action.SCHEDULED;
+                }
+                return Action.SUCCEEDED;
+            }
+        };
+        writer.iterate();
+    }
+
+    private static class CountingByteBufferAccumulator
```

Comment:
I don't like the name.  What does "Counting" mean?     Also it is aggregating not accumulating.   I think `ByteBufferAggregator` would be better.
Also should be in `io` or `util`

Reply:
Done.